### PR TITLE
feat: Support aggregation functions in Python

### PR DESCRIPTION
**Pull request by bjchambers**
Introduces Python classes for representing windows and flattens them into arguments for the corresponding aggregation functions.

This also changes the AST to DFG logic so that it only tries to flatten if the original AST is available. This does mean that ticks won't be recreated, meaning they won't have the proper domain.

Both of these likely need to be extended to cover other aggregations.

### CODE REVIEW

File: crates/sparrow-compiler/src/functions/aggregation.rs [hunk 1]
```diff
@@ -39,7 +39,7 @@ pub(super) fn register(registry: &mut Registry) {
     registry
         .register("sum<N: number>(input: N, window: window = null) -> N ")
         .with_dfg_signature(
-            "sum<N: number>(input: N, window: window = null, duration: i64 = null) -> N",
+            "sum<N: number>(input: N, window: bool = null, duration: i64 = null) -> N",
```
Comment (jordanrfrazier) [inline crates/sparrow-compiler/src/functions/aggregation.rs:42]: Looking into this. This will cause sparrow tests to fail.

Comment (jordanrfrazier) [reply crates/sparrow-compiler/src/functions/aggregation.rs:42]: This type here is a`window` (and not `boolean`) for type-checking the dfg. The dfg checks that the window argument to the signature is in fact a window, and not just a boolean. After if verifies that and instantiates the types, it flattens the window argument. Additionally, the dfg signature in the registry needs to have the duration argument, so it can include the duration within the pattern.
```
        .with_dfg_signature("sum<N: number>(input: N, window: window = null, duration: i64 = null) -> N")
        .with_implementation(Implementation::new_pattern(&format!(
            "(sum ({}) ({}) ({}))",
            "transform (if ?input_is_new ?input_value) (merge_join ?input_op ?window_op)",
            "?window_value",
            "?duration_value"
```

File: crates/sparrow-compiler/src/ast_to_dfg.rs [hunk 2]
```diff
@@ -517,15 +517,28 @@ pub fn add_to_dfg(
...
+                        dfg.exit_env();
+                        result
+                    }
+                    None => {
+                        // If `expr` is None, we're running the Python builder code,
+                        // which already flattened things.
+                        //
+                        // Note that this won't define the `condition_input` for the
+                        // purposes of ticks.
+                        (args[1].clone(), args[2].clone())
```
Comment (jordanrfrazier) [inline crates/sparrow-compiler/src/ast_to_dfg.rs:538]: The difference here is that `args[1]` is already a boolean, not a window. That's why the change to the `dfg_signature` works with the python test, and not sparrow. 

Which is somewhat interesting -- with some refactoring to the `ast_to_dfg`, we could probably support `window=daily()` instead of `window=since(daily())`.

Comment (jordanrfrazier) [reply crates/sparrow-compiler/src/ast_to_dfg.rs:538]: This should work after fixing my signature bugs in https://github.com/kaskada-ai/kaskada/pull/587

File: crates/sparrow-compiler/src/functions/aggregation.rs (hunk 1 above)
Comment (jordanrfrazier) [reply crates/sparrow-compiler/src/functions/aggregation.rs:42]: This should work after fixing my signature bugs in https://github.com/kaskada-ai/kaskada/pull/587

File: sparrow-py/pysrc/sparrow_py/expr.py [hunk 3]
```diff
@@ -285,3 +299,12 @@ def run_to_csv_string(self) -> str:
         """Execute the expression and return the CSV as a string."""
         # Convert PyArrow -> Pandas -> CSV, omitting the index (row).
         return self.run().to_csv(index=False)
+
+def _aggregation(op: str, input: Expr, window: Optional[Window]) -> Expr:
+    """Helper for creating an aggregation."""
+    if window is None:
+       return Expr.call(op, input, None, None)
+    elif isinstance(window, SinceWindow):
+       return Expr.call(op, input, window._predicate, 1)
```
Comment (jordanrfrazier) [inline sparrow-py/pysrc/sparrow_py/expr.py:308]: Can you change the `1` to `None`? This will ensure we actually use the `since` window implementations, which should be more performant than the TwoStacks. (Also, collect only supports `since`)

Ideally we pass the type of the window rather than relying on "number of non-null args" to determine window type..but, future change.

File: sparrow-py/src/expr.rs [hunk 4]
```diff
@@ -31,9 +36,16 @@ impl Expr {
...
-                arg.into_ast_dfg_ref(&mut rust_session)
-                    // DO NOT SUBMIT: Better error handling.
-                    .map_err(|_| PyRuntimeError::new_err("unable to create literal"))
+                match arg {
+                    None => rust_session
+                        .add_literal(Literal::Null)
+                        .map_err(|_| PyRuntimeError::new_err("unable to create null literal")),
+                    Some(arg) => {
+                        arg.into_ast_dfg_ref(&mut rust_session)
+                            // DO NOT SUBMIT: Better error handling.
```
Comment (jordanrfrazier) [inline sparrow-py/src/expr.rs:45]: blocker?

**Review by jordanrfrazier — APPROVED** [review]

File: sparrow-py/src/expr.rs (hunk 4 above)
Comment (bjchambers) [reply sparrow-py/src/expr.rs:45]: not really -- definitely need to really improve the error story, but TBD
